### DISCUSSION TITLE
Applying formatter to virtual document

### DISCUSSION THREAD

**Discussion by AllanJard** [microsoft/vscode-discussions#609] [Extension Development QnA] [answered]
I'm trying to work out how to apply a formatter to a virtual document, and not really getting anywhere. I can create a virtual document using `registerTextDocumentContentProvider` to register a handler, and then construct it with:

```js
	const originalUri = document.uri.toString(true);
	const vdocUriString = `embedded-content://css/${encodeURIComponent(
		originalUri
	)}.css`;
	const vdocUri = Uri.parse(vdocUriString);

	virtualDocumentContents.set(originalUri, matches[1]);
```

That lets me call the formatter for the content:

```js
	const res = await commands.executeCommand<TextEdit[]>(
		'vscode.executeFormatDocumentProvider',
		vdocUri,
		{
			tabSize: 2,
			insertSpaces: true
		}
	);
```

Inspecting the result I can see the `TextEdit[]` entry contains the replacement that I want (with a range, so I can't just do a string replacement).

I presume that I need to apply those TextEdits to my virtual document, but I have no idea how to:

1. Actually get the `document` (I've got the Url, but how do I get the `TextDocument` object from that)
2. How to apply the `TextEdit` array to the document. I can't see a suitable method - so maybe I'm barking up the wrong tree!

Any thoughts on how to do this?

**Comment by isidorn** [comment]
@allanjard Is this still the issue? Did you figure out the answer to this question? Thanks

**Comment by AllanJard** [comment]
I didn't I'm afraid. I've set that work aside as I wasn't able to get any further with it. If you have any suggestions, they would be warmly welcomed.

**Reply by isidorn** [reply]
Maybe @jrieken can help

**Reply by jrieken** [reply, marked as answer]
Formatting virtual documents isn't supported. They are held as readonly documents and hence formatting isn't available for them

**Comment by AllanJard** [comment]
Thank you - that clarifies why I was having some issues with it then :-).
